### PR TITLE
Allow more leniency when defining the server url

### LINKED ISSUE #4144: More flexible Realm Urls for SyncConfiguration
Right now we require Realm Urls to `new SyncConfiguration.Builder(user, url)` to have the exact format:

```
<realm|realms>://<baseUrl>/<path>
```

This is needlessly restrictive and we can relax it the following ways:

1) Support only `<path>` ... baseUrl/scheme will be inferred from the Users auth server
2) Support using `<realm|realms>://<baseUrl>/<path>` -> Automatically replace http/https with realm/realms

### PR DESCRIPTION
**Pull request by cmelchior**
Fixes #4144 

@realm/java

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 ## 2.3.1
 
+### Enhancements
+
+* The `serverUrl` given to `SyncConfiguration.Builder()` is now more lenient and will also accept only paths as argument (#).
```
Comment (nhachicha) [inline CHANGELOG.md]: `#4144`

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 2]
```diff
@@ -277,7 +277,10 @@ boolean isSyncConfiguration() {
          *
          * @param user the user for this Realm. An authenticated {@link SyncUser} is required to open any Realm managed
          *             by a Realm Object Server.
-         * @param uri URI identifying the Realm.
+         * @param uri URI identifying the Realm. If only a path like {@code /~/default} is given, the configuration will
+         *            assume the file is located on the same server returned by {@link SyncUser#getAuthenticationUrl()}.
+         *            the user server used
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: the authentication server used to ...?

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java [hunk 3]
```diff
@@ -116,6 +109,39 @@ public void serverUrl_setsFolderAndFileName() {
...
+                { createTestUser("http://ros.realm.io/auth"), "ros.realm.io/~/default", "realm://ros.realm.io/~/default" },
+                { createTestUser("http://ros.realm.io/auth"), "//ros.realm.io/~/default", "realm://ros.realm.io/~/default" },
+                { createTestUser("https://ros.realm.io/auth"), "ros.realm.io/~/default", "realms://ros.realm.io/~/default" },
+                { createTestUser("https://ros.realm.io/auth"), "//ros.realm.io/~/default", "realms://ros.realm.io/~/default" },
+
+                // Automatically replace http|https with realm|realms
+                { createTestUser(), "http://ros.realm.io/~/default", "realm://ros.realm.io/~/default" },
+                { createTestUser(), "https://ros.realm.io/~/default", "realms://ros.realm.io/~/default" }
+        };
+
```
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java]: What about use cases where the user specify the [custom|default] authentication port will you skip the port?
Ex:
```java
 { createTestUser("http://ros.realm.io:7777/auth"), "ros.realm.io/~/default", "realm://ros.realm.io/~/default" }
```
?

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 4]
```diff
@@ -314,15 +317,49 @@ private void validateAndSet(String uri) {
             }
 
             try {
-                serverUrl = new URI(uri);
+                serverUrl = new URI(uri.toLowerCase());
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: why `toLowerCase()`? URI should be case sensitive, at least I guess Apache & nginx are

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java (hunk 3 above)
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java]: also would love to see a test using ip address.

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 5]
```diff
@@ -314,15 +317,49 @@ private void validateAndSet(String uri) {
...
+                String path = serverUrl.getPath();
+                if (path != null && !path.startsWith("/")) {
+                    path = "/" + path;
+                }
+
+                serverUrl = new URI(serverScheme,
+                        serverUrl.getUserInfo(),
+                        host,
+                        serverUrl.getPort(),
+                        path.replace(host + "/", ""), // Remove host if it accidentially was interpreted as a path segment
```
Comment (kneth) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: What if the Realm has the same name as the host?

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java (hunk 3 above)
Comment (cmelchior) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java]: Yes, we should skip any custom port from the auth URL. Adding a test for it. Also adding a test using IP addresses

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java (hunk 5 above)
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: Added unit test for this as well.

**Pull request merged**